// @shopify/polaris/build/cjs/components/ActionList/components/SearchField/SearchField.js
'use strict';

var polarisIcons = require('@shopify/polaris-icons');
var React = require('react');
var SearchField$1 = require('./SearchField.scss.js');
var hooks = require('../../../../utilities/i18n/hooks.js');
var Icon = require('../../../Icon/Icon.js');
var Text = require('../../../Text/Text.js');

function SearchField({
  value,
  placeholder,
  onChange
}) {
  const i18n = hooks.useI18n();
  const input = React.useRef(null);
  const searchId = React.useId();
  const handleChange = React.useCallback(({
    currentTarget
  }) => {
    onChange(currentTarget.value);
  }, [onChange]);
  const handleClear = React.useCallback(() => {
    if (!input.current) {
      return;
    }
    input.current.value = '';
    onChange('');
    input.current.focus();
  }, [onChange]);
  const clearMarkup = value !== '' && /*#__PURE__*/React.createElement("button", {
    type: "button",
    "aria-label": i18n.translate('Polaris.ActionList.SearchField.clearButtonLabel'),
    className: SearchField$1.default.Clear,
    onClick: handleClear,
    onBlur: () => {
      handleClear();
    }
  }, /*#__PURE__*/React.createElement(Icon.Icon, {
    source: polarisIcons.CircleCancelMinor
  }));
  const handleRef = ref => {
    input.current = ref;

    // It won't focus if it's on the same tick as when it renders
    setTimeout(() => {
      ref?.focus();
    });
  };
  return /*#__PURE__*/React.createElement("div", {
    className: SearchField$1.default.SearchField
  }, /*#__PURE__*/React.createElement(Text.Text, {
    as: "span",
    visuallyHidden: true
  }, /*#__PURE__*/React.createElement("label", {
    htmlFor: searchId
  }, i18n.translate('Polaris.ActionList.SearchField.search'))), /*#__PURE__*/React.createElement("input", {
    id: searchId,
    className: SearchField$1.default.Input,
    placeholder: placeholder,
    type: "search",
    autoCapitalize: "off",
    autoComplete: "off",
    autoCorrect: "off",
    ref: handleRef,
    value: value,
    onChange: handleChange,
    onKeyDown: preventDefault
  }), /*#__PURE__*/React.createElement("span", {
    className: SearchField$1.default.Icon
  }, /*#__PURE__*/React.createElement(Icon.Icon, {
    source: polarisIcons.SearchMinor
  })), clearMarkup, /*#__PURE__*/React.createElement("div", {
    className: SearchField$1.default.Backdrop
  }));
}
function preventDefault(event) {
  if (event.key === 'Enter') {
    event.preventDefault();
  }
}

exports.SearchField = SearchField;
